README.txt
Welcome to HDBC-MissingH.

HDBC-MissingH is a package to integrate HDBC (Haskell Database
Connectivity) with the MissingH library.

At the moment, this means that any HDBC database can act as a backend
for the MissingH.AnyDBM interface.

Please see doc/MissingH-AnyDBM-HDBCDBM.html for more information.

PREREQUISITES
-------------

1) You'll need either GHC 6.4.1 or above, or Hugs 2005xx or above.

2) Also, you need MissingH 0.13.0 or above.  Download it from
   http://quux.org/devel/missingh

3) Finally, you need HDBC 0.99.0 or above.  Download it from
   http://quux.org/devel/hdbc

INSTALLATION
------------

The steps to install are:

1) ghc --make -o setup Setup.lhs

2) ./setup configure

3) ./setup build

4) ./setup install   (as root)

If you're on Windows, you can omit the leading "./".

Documentation is in doc/.

USAGE
-----

To use with hugs, you'll want to use hugs -98.

To use with GHC, you'll want to use -package HDBC-MissingH in your programs.
Or, with Cabal, use Build-Depends: HDBC-MissingH.

-- John Goerzen
   December 2005


debian/changelog
hdbc-missingh (1.0.1.1) unstable; urgency=low

  * Add missing build-dep on libghc6-hunit-dev.

 -- John Goerzen <jgoerzen@complete.org>  Sat, 21 Oct 2006 20:15:32 -0500

hdbc-missingh (1.0.1.0) unstable; urgency=low

  * Rebuilding for GHC 6.6.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 19 Oct 2006 21:52:53 -0500

hdbc-missingh (1.0.0.1) unstable; urgency=low

  * Rebuilding for HDBC 1.0.1

 -- John Goerzen <jgoerzen@complete.org>  Sat, 22 Jul 2006 07:29:15 -0500

hdbc-missingh (1.0.0.0) unstable; urgency=low

  * Blessing for 1.0.0.
  * Rebuilding for Debian with GHC 6.4.2.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 29 Jun 2006 18:50:18 -0500

hdbc-missingh (0.99.2.3) unstable; urgency=low

  * Rebuilding for MissingH 0.14.2.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 13 Apr 2006 00:40:51 -0500

hdbc-missingh (0.99.2.2) unstable; urgency=low

  * Fix typo in control.  Closes: #361170.

 -- John Goerzen <jgoerzen@complete.org>  Fri,  7 Apr 2006 06:22:25 -0500

hdbc-missingh (0.99.2.1) unstable; urgency=low

  * Rebuilding for new MissingH.

 -- John Goerzen <jgoerzen@complete.org>  Thu,  6 Apr 2006 02:47:37 -0500

hdbc-missingh (0.99.2.0) unstable; urgency=low

  * Rebuild against HDBC 0.99.2.
  * Don't try to commit against DBs that don't support transactions.

 -- John Goerzen <jgoerzen@complete.org>  Tue, 28 Mar 2006 05:46:12 -0600

hdbc-missingh (0.99.1.0) unstable; urgency=low

  * Rebuilt against HDBC 0.99.1.

 -- John Goerzen <jgoerzen@complete.org>  Mon, 27 Mar 2006 22:58:39 -0600

hdbc-missingh (0.99.0.1) unstable; urgency=low

  * Updated build-deps on MissingH.

 -- John Goerzen <jgoerzen@complete.org>  Mon, 20 Mar 2006 20:39:34 -0600

hdbc-missingh (0.99.0.0) unstable; urgency=low

  * Initial release.

 -- John Goerzen <jgoerzen@complete.org>  Thu, 29 Dec 2005 15:24:13 -0600


testsrc/Tests.hs
{- arch-tag: Tests main file
Copyright (C) 2004 John Goerzen <jgoerzen@complete.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
-}

module Tests(tests) where
import Test.HUnit
import qualified AnyDBMtest

test1 = TestCase ("x" @=? "x")

tests = TestList [TestLabel "test1" test1,
                 TestLabel "AnyDBM" AnyDBMtest.tests]



MissingH/AnyDBM/HDBCDBM.hs
{-
Copyright (C) 2005 John Goerzen <jgoerzen@complete.org>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

-}

{- |
   Module     : MissingH.AnyDBM.HDBCDBM
   Copyright  : Copyright (C) 2005 John Goerzen
   License    : GNU LGPL, version 2.1 or above

   Maintainer : John Goerzen <jgoerzen@complete.org>
   Stability  : provisional
   Portability: portable

HDBC backend for DBM databases

Written by John Goerzen, jgoerzen\@complete.org

-}

module MissingH.AnyDBM.HDBCDBM(HDBCDBM, openSimpleHDBCDBM,
                               openHDBCDBM) where

import MissingH.AnyDBM
import Database.HDBC
import System.IO(IOMode)
import Data.Char
import Control.Monad
import Control.Exception 
import Control.Concurrent.MVar

{- | Main HDBC DBM type.

Please note that accessor methods on these objects will not return SqlErrors
since AnyDBM users won\'t expect them.  Instead, they will marshal those
into IOErrors.

Also, please note that you cannot expect a HDBCDBM to be able to store
the NUL character since most SQL databases can\'t store it.  If you need to be
able to store that character with your data, I suggest passing your data
through read\/show around this module.
-}
data HDBCDBM = HDBCDBM {conn :: Connection,
                        tablename :: String,
                        keycolname :: String,
                        valcolname :: String,
                        lock :: MVar ()}

{- | Opens a DBM connection to the specified table and database.

If the given table does not already exist, it will be created for you
automatically.

The column names @dbmkey@ and @dbmval@ will be used.

The 'Database.HDBC.clone' method will be used to obtain a new handle to 
the database for internal use.  That will help prevent conflicts with things
like commits from occuring with the main application handle.  You should not
have to worry about side-effects to your main application as a result of
using this function.

If you need more control, see 'openHDBCDBM'

Please note that HDBCDBM checks dbTransactionSupport.  If your database
indicates that transactions are /NOT/ supported (such as MySQL), then
atomicity and safety may not be guaranteed in multithreaded programs. -}
openSimpleHDBCDBM :: String -> Connection -> IO HDBCDBM
openSimpleHDBCDBM itablename iconn =
    do mydbh <- clone iconn
       tablelist <- getTables mydbh
       when (not ((map toLower) itablename `elem` tablelist))
            (withTransaction mydbh (createtable (map toLower itablename)))
       newmv <- newMVar ()
       return $ HDBCDBM {conn = mydbh,
                         tablename = (map toLower) itablename,
                         keycolname = "dbmkey",
                         valcolname = "dbmval",
                         lock = newmv}
    where createtable tablename dbh =
              run dbh ("CREATE TABLE " ++ tablename ++ 
                        "(dbmkey text NOT NULL PRIMARY KEY, " ++
                        "dbmval text NOT NULL)") [] >> return ()

{- | Opens a DBM connection to the specified table, column name for key,
column name for value, and database.

It is an error if the given table does not exist, lacks the specified columns,
or has an incorrect type definition for the specified columns.  However,
it is not guaranteed that this error will be detected at object creation time.

This function does not create a table for you if it doesn't exist.

You can use the table created and used by 'openSimpleHDBCDBM' like this:

>openHDBCDBM "tablename" "dbmkey" "dbmval" conn

Incidentally, such a table could be created with:

>CREATE TABLE tablename (dbmkey text NOT NULL PRIMARY KEY, dmval text NOT NULL)

The 'Database.HDBC.clone' method will be used to obtain a new handle to 
the database for internal use.  That will help prevent conflicts with things
like commits from occuring with the main application handle.  You should not
have to worry about side-effects to your main application as a result of
using this function. -}
openHDBCDBM :: String           -- ^ Name of table to use
            -> String           -- ^ Name of column with keys
            -> String           -- ^ Name of column with values
            -> Connection       -- ^ HDBC connection object to clone
            -> IO HDBCDBM       -- ^ Returned DBM object
openHDBCDBM itablename ikeyname ivalname iconn =
    do mydbh <- clone iconn
       tablelist <- getTables mydbh
       let mytablename = map toLower itablename
       when (not (mytablename `elem` tablelist))
            (fail $ "Table " ++ mytablename ++ " not in database.")
       newmv <- newMVar ()
       return $ HDBCDBM {conn = mydbh,
                         tablename = mytablename,
                         keycolname = ikeyname,
                         valcolname = ivalname,
                         lock = newmv}

basequery dbm = " FROM " ++ (tablename dbm) ++
               " WHERE " ++ (keycolname dbm) ++ " = ?"

querykey :: HDBCDBM -> String
querykey dbm = "SELECT " ++ (keycolname dbm) ++ ", " ++
               (valcolname dbm) ++ basequery dbm

deletequery :: HDBCDBM -> String
deletequery dbm = "DELETE " ++ basequery dbm

insertquery dbm = "INSERT INTO " ++ (tablename dbm) ++
                  " (" ++ (keycolname dbm) ++ ", " ++ (valcolname dbm) ++ 
                  ") VALUES (?, ?)"

updatequery dbm = "UPDATE " ++ (tablename dbm) ++ " SET " ++ 
                  (valcolname dbm) ++ " = ? WHERE " ++ (keycolname dbm) ++
                  " = ?"
        
withlock dbm f = withMVar (lock dbm) (\_ -> f)

instance AnyDBM HDBCDBM where
    closeA dbm = handleerror dbm $ 
                 do condcommit (conn dbm)
                    disconnect (conn dbm)

    flushA dbm = handleerror dbm $ withlock dbm $ condcommit (conn dbm)

    insertA dbm key value = handleerror dbm $ withlock dbm $
            do count <- run (conn dbm) (updatequery dbm) 
                        [toSql value, toSql key]
               case count of
                 0 -> -- No change, need to insert it.
                      do run (conn dbm) (insertquery dbm) 
                          [toSql key, toSql value]
                         return ()
                 1 -> return () -- We tweaked 1 row
                 x -> fail $ "HDBC insertA: unexpected number of rows updated: " ++ show x

    deleteA dbm key = handleerror dbm $ withlock dbm $
            run (conn dbm) (deletequery dbm) [toSql key] >> return ()

    lookupA dbm key = handleerror dbm $ withlock dbm $
        do res <- quickQuery (conn dbm) (querykey dbm) [toSql key]
           case res of
             [] -> return Nothing
             [[_, value]] -> return (Just (fromSql value))
             x -> fail $ "lookupA: unexpected return value " ++ show x

    toListA dbm = handleerror dbm $ withlock dbm $
        do res <- quickQuery (conn dbm) 
                  ("SELECT " ++ keycolname dbm ++ ", " ++ valcolname dbm ++
                   " FROM " ++ tablename dbm) []
           -- This forces the entire result to be read before returning
           -- any of it.  Ugly for now -- AnyDBM API changes could improve
           -- this.
           return $ seq (length res) map convrow res
        where convrow [k, v] = (fromSql k, fromSql v)
              convrow x = error $ "toListA: unexpected row " ++ show x

{- When a SQL error occurs, a transaction can be left in an indeterminate
state.  Commit things, then re-raise the error.

Also, wrap it all in handleSqlError since AnyDBM users might not expect
a SqlError. -}
handleerror dbm action = handleSqlError $ catchSql action handler
    where handler e =
              do condcommit (conn dbm)
                 throwDyn e

{- Only do a commit if the DB supports it. -}
condcommit dbh =
    when (dbTransactionSupport dbh) (commit dbh)


testsrc/AnyDBMtest.hs
{- arch-tag: AnyDBM tests main file
Copyright (C) 2004-2005 John Goerzen <jgoerzen@complete.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
-}

module AnyDBMtest(mf, generic_persist_test, generic_test, tests) where
import Test.HUnit
import MissingH.List
import MissingH.IO.HVFS
import MissingH.IO.HVFS.InstanceHelpers
import MissingH.AnyDBM
import MissingH.AnyDBM.StringDBM
import MissingH.AnyDBM.FiniteMapDBM
import MissingH.AnyDBM.MapDBM
import MissingH.AnyDBM.HDBCDBM
import Database.HDBC.Sqlite3
import Database.HDBC.PostgreSQL
import System.Directory
import MissingH.IO.HVFS.Utils
import MissingH.Path.FilePath
import Data.HashTable
import Data.List(sort)
import Control.Exception(finally)

mf :: AnyDBM a => IO b -> (b -> IO a) -> String -> (a -> Assertion) -> Test
mf initfunc openfunc msg code =
    TestLabel msg $ TestCase $ do i <- initfunc
                                  h <- openfunc i
                                  finally (code h) (closeA h)
        
infix 1 @>=?
(@>=?) :: (Eq a, Show a) => a -> IO a -> Assertion
(@>=?) exp res = do r <- res
                    exp @=? r

deleteall h = do k <- keysA h
                 mapM_ (deleteA h) k
                 [] @>=? keysA h

weirdl = sort $ [("", "empty"), 
                 ("foo\nbar", "v1\0v2"),
                 ("v3\x01v4", ""),
                 ("k\0ey", "\xFF")]

weirdl2 = sort $ [("", "empty"),
                  ("foo\nbar", "v1\x01v2"),
                  ("v3\x02v4", ""),
                  ("key", "\xFF")]

createdir = TestCase $ createDirectory "testtmp"
removedir = TestCase $ recursiveRemove SystemFS "testtmp"

generic_test wl initfunc openfunc =
    let f = mf initfunc openfunc in
        [
         createdir
        ,f "empty" $ \h -> do [] @>=? keysA h
                              [] @>=? valuesA h
                              [] @>=? toListA h
                              Nothing @>=? lookupA h "foo"
                     
        ,f "basic" $ \h -> do insertA h "key" "value"
                              (Just "value") @>=? lookupA h "key"
                              [("key", "value")] @>=? toListA h
                              insertA h "key" "v2"
                              [("key", "v2")] @>=? toListA h
                              deleteA h "key"
                              [] @>=? toListA h
        ,f "mult" $ \h -> do insertListA h [("1", "2"), ("3", "4"), ("5", "6")]
                             [("1", "2"), ("3", "4"), ("5", "6")] @>=? 
                                (toListA h >>= return . sort)
                             ["1", "3", "5"] @>=? (keysA h >>= return . sort)
                             ["2", "4", "6"] @>=? (valuesA h >>= return . sort)
                             deleteall h
        ,f "weirdchars" $ \h -> do insertListA h wl
                                   wl @>=? (toListA h >>= return . sort)
                                   deleteall h
        ,removedir
        ]

generic_persist_test wl initfunc openfunc  =
    let f = mf initfunc openfunc in
        [
         createdir
        ,f "empty" deleteall 
        ,f "weirdpop" $ \h -> insertListA h wl
        ,f "weirdcheck" $ \h -> do wl @>=? (toListA h >>= return . sort)
                                   deleteall h
                                   insertA h "key" "value"
        ,f "step3" $ \h -> do [("key", "value")] @>=? (toListA h >>= return . sort)
                              insertA h "key" "v2"
                              insertA h "z" "y"
        ,f "step4" $ \h -> do [("key", "v2"), ("z", "y")] @>=?
                                 (toListA h >>= return . sort)
        ,f "cleanupdb" deleteall
        ,removedir
        ]

test_hashtable = generic_test weirdl (return ())
                  (\_ -> ((new (==) hashString)::IO (HashTable String String)))

test_finitemap = generic_test weirdl (return ())
                  (\_ -> newFiniteMapDBM)
test_mapdbm = generic_test weirdl (return ())
                  (\_ -> newMapDBM)
test_stringdbm = generic_persist_test weirdl (return SystemFS)
                   (\f -> openStringVDBM f (joinPaths "testtmp" "StringDBM") ReadWriteMode)
                 ++
                 generic_test weirdl (return SystemFS)
                   (\f -> openStringVDBM f (joinPaths "testtmp" "StringDBM") ReadWriteMode)

test_hdbcdbm cf = generic_persist_test weirdl2 (cf)
                  (\f -> openSimpleHDBCDBM "hdbcdbmtest" f)
               ++
               generic_test weirdl2 (cf)
                  (\f -> openSimpleHDBCDBM "hdbcdbmtest" f)
    where connsql3 = connectSqlite3 (joinPaths "testtmp" "HDBCDBM")

tests = TestList [TestLabel "HashTable" (TestList test_hashtable),
                  TestLabel "StringDBM" (TestList test_stringdbm),
                  TestLabel "FiniteMap" (TestList test_finitemap),
                  TestLabel "MapDBM" (TestList test_mapdbm),
                  TestLabel "HDBCDBM-sqllite3" 
                                (TestList (test_hdbcdbm (connectSqlite3 (joinPaths "testtmp" "HDBCDBM")))),
                  TestLabel "HDBCDBM-psql" (TestList (test_hdbcdbm $ connectPostgreSQL ""))
                 ]
